File: StripMate/Core/ViewModels/SupportChatViewModel.swift
import Foundation
import FirebaseFirestore
import FirebaseAuth

@MainActor @Observable
final class SupportChatViewModel {

    // MARK: - Properties

    var messages: [SupportMessage] = []
    var inputText: String = ""
    var isLoading: Bool = false

    /// Stored in an `IsolatedRef` so the nonisolated `deinit` can remove the
    /// Firestore listener without `nonisolated(unsafe)`.
    private let listener = IsolatedRef<ListenerRegistration?>(nil)
    private let db = Firestore.firestore()

    deinit {
        listener.value?.remove()
    }

    var currentUserId: String? {
        Auth.auth().currentUser?.uid
    }

    // MARK: - Listener

    func listenToMessages() {
        guard let uid = currentUserId else { return }
        isLoading = true

        listener.value = db
            .collection("support_chats")
            .document(uid)
            .collection("messages")
            .order(by: "timestamp", descending: false)
            .addSnapshotListener { [weak self] snapshot, error in
                guard let self else { return }
                self.isLoading = false

                guard let documents = snapshot?.documents else { return }
                self.messages = documents.compactMap { doc in
                    let data = doc.data()
                    guard let senderId = data["senderId"] as? String,
                          let text = data["text"] as? String,
                          let ts = data["timestamp"] as? Timestamp,
                          let isAdmin = data["isAdmin"] as? Bool else { return nil }

                    let readAt = (data["readAt"] as? Timestamp)?.dateValue()

                    return SupportMessage(
                        id: doc.documentID,
                        senderId: senderId,
                        text: text,
                        timestamp: ts.dateValue(),
                        isAdmin: isAdmin,
                        readAt: readAt
                    )
                }
            }
    }

    // MARK: - Send

    func sendMessage() async {
        let trimmed = inputText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let uid = currentUserId else { return }

        let text = trimmed
        inputText = ""

        let docRef = db
            .collection("support_chats")
            .document(uid)
            .collection("messages")
            .document()

        let payload: [String: Any] = [
            "senderId": uid,
            "text": text,
            "timestamp": FieldValue.serverTimestamp(),
            "isAdmin": false
        ]

        // Parent dokümanı oluştur (yoksa) — admin app thread'i bulabilsin
        let parentRef = db.collection("support_chats").document(uid)
        do {
            try await parentRef.setData(["createdAt": FieldValue.serverTimestamp(), "userId": uid], merge: true)
        } catch {
            AppLogger.service.debug("\(error)")
        }

        do {
            try await docRef.setData(payload)
        } catch {
            AppLogger.service.debug("\(error)")
        }
    }

    // MARK: - Cleanup

    func stopListening() {
        listener.value?.remove()
        listener.value = nil
    }
}

File: android/app/src/main/java/com/celalbasaran/stripmate/ui/screen/support/SupportChatViewModel.kt
package com.celalbasaran.stripmate.ui.screen.support

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.celalbasaran.stripmate.service.auth.AuthRepository
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.util.Date
import javax.inject.Inject

data class SupportMessage(
    val id: String = "",
    val senderId: String = "",
    val text: String = "",
    val timestamp: Date = Date(),
    val isAdmin: Boolean = false
)

@HiltViewModel
class SupportChatViewModel @Inject constructor(
    private val authRepository: AuthRepository
) : ViewModel() {

    private val db = FirebaseFirestore.getInstance()

    private val _messages = MutableStateFlow<List<SupportMessage>>(emptyList())
    val messages: StateFlow<List<SupportMessage>> = _messages.asStateFlow()

    private val _inputText = MutableStateFlow("")
    val inputText: StateFlow<String> = _inputText.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private var listener: ListenerRegistration? = null

    private val currentUserId: String?
        get() = authRepository.currentUserId()

    init {
        listenToMessages()
    }

    private fun listenToMessages() {
        val uid = currentUserId ?: return
        _isLoading.value = true

        listener = db
            .collection("support_chats")
            .document(uid)
            .collection("messages")
            .orderBy("timestamp", Query.Direction.ASCENDING)
            .addSnapshotListener { snapshot, error ->
                _isLoading.value = false
                if (error != null || snapshot == null) return@addSnapshotListener

                _messages.value = snapshot.documents.mapNotNull { doc ->
                    val data = doc.data ?: return@mapNotNull null
                    val senderId = data["senderId"] as? String ?: return@mapNotNull null
                    val text = data["text"] as? String ?: return@mapNotNull null
                    val isAdmin = data["isAdmin"] as? Boolean ?: false
                    val timestamp = (data["timestamp"] as? com.google.firebase.Timestamp)?.toDate() ?: Date()

                    SupportMessage(
                        id = doc.id,
                        senderId = senderId,
                        text = text,
                        timestamp = timestamp,
                        isAdmin = isAdmin
                    )
                }
            }
    }

    fun updateInput(text: String) {
        _inputText.value = text
    }

    fun sendMessage() {
        val text = _inputText.value.trim()
        if (text.isBlank()) return
        val uid = currentUserId ?: return

        _inputText.value = ""

        viewModelScope.launch {
            try {
                // Ensure parent document exists so admin app can find the thread
                val parentRef = db.collection("support_chats").document(uid)
                parentRef.set(
                    mapOf(
                        "createdAt" to FieldValue.serverTimestamp(),
                        "userId" to uid
                    ),
                    com.google.firebase.firestore.SetOptions.merge()
                ).await()

                // Add message
                val payload = hashMapOf<String, Any>(
                    "senderId" to uid,
                    "text" to text,
                    "timestamp" to FieldValue.serverTimestamp(),
                    "isAdmin" to false
                )
                parentRef.collection("messages").add(payload).await()
            } catch (_: Exception) {
                // Silently fail - message will appear when Firestore syncs
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        listener?.remove()
        listener = null
    }
}
